specs/networking/p2p-interface.md

```diff
@@ -0,0 +1,743 @@
+
+Pubsub is a technique to broadcast/disseminate data across a network rapidly. Such data is packaged in fire-and-forget messages that do not require a response from every recipient. Peers subscribed to a topic participate in the propagation of messages in that topic.
+
+The alternative is to maintain a fully connected mesh (all peers connected to each other 1:1), which scales poorly (O(n^2)).
+
+### Why are we using topics to segregate encodings, yet only support one encoding?
+
+For future extensibility with almost zero overhead now (besides the extra bytes in the topic name).
+
+### How do we upgrade gossip channels (e.g. changes in encoding, compression)?
+
+Such upgrades lead to fragmentation, so they’ll need to be carried out in a coordinated manner most likely during a hard fork.
+
+### Why are the topics strings and not hashes?
+
+Topics names have a hierarchical structure. In the future, gossipsub may support wildcard subscriptions (e.g. subscribe to all children topics under a root prefix) by way of prefix matching. Enforcing hashes for topic names would preclude us from leveraging such features going forward.
+
+No security or privacy guarantees are lost as a result of choosing plaintext topic names, since the domain is finite anyway, and calculating a digest's preimage would be trivial.
+
+Furthermore, the ETH2 topic names are shorter than their digest equivalents (assuming SHA-256 hash), so hashing topics would bloat messages unnecessarily.
+
+### Why are there `SHARD_SUBNET_COUNT` subnets, and why is this not defined?
+
+Depending on the number of validators, it may be more efficient to group shard subnets and might provide better stability for the gossipsub channel. The exact grouping will be dependent on more involved network tests. This constant allows for more flexibility in setting up the network topology for attestation aggregation (as aggregation should happen on each subnet).
+
+### Why are we sending entire objects in the pubsub and not just hashes?
+
+Entire objects should be sent to get the greatest propagation speeds. If only hashes are sent, then block and attestation propagation is dependent on recursive requests from each peer. In a hash-only scenario, peers could receive hashes without knowing who to download the actual contents from. Sending entire objects ensures that they get propagated through the entire network.
+
+### Should clients gossip blocks if they *cannot* validate the proposer signature due to not yet being synced, not knowing the head block, etc?
+
+The prohibition of unverified-block-gossiping extends to nodes that cannot verify a signature due to not being fully synced to ensure that such (amplified) DOS attacks are not possible.
+
+### How are we going to discover peers in a gossipsub topic?
+
+Via discv5 topics. ENRs should not be used for this purpose, as they store identity, location and capability info, not volatile advertisements.
```

Comment:
It would be useful to add a link to Topic Advertisement here.

Reply:
I think it makes sense to have discv5 section here that outlines how eth2.0 will use it specifically.

Reply:
link added, marked as under construction as discv5 specifics are still being worked out

514ff5814f233836eba8b45d9bb5d94b97d7a60e